README.txt
.. contents::

- Code repository: https://ingeniweb.svn.sourceforge.net/svnroot/ingeniweb/plone.recipe.pound/trunk


CHANGES.txt
0.5.6-eea - (23-07-2013)
========================

  - added log_file option to allow output redirection to this file,
    when using with log_facility set to -
    [Tiberiu Ichim]
  - only set ID in session block of pound.cfg if value of sessiontype is 
    COOKIE; otherwise Pound will not start (e.g., ID is set but type is IP)
    [bryanw]
  - Delete script before writing to it, this way we avoid chmod permission errors
    when the current user is not the script owner.
    [hman]
  - Markup fixes and cleanup
    [alecghica]
  - Fixed poundctl.tpl to make logs file persistent
    [alecghica]

0.5.5 (2009-06-02)
==================

  - cmmi change extra_options configuration
    [youenn]

0.5.4 (2008-12-12)
==================

  - Add new options to configure sticky session
    [youenn]


0.5.3 (2008-10-23)
==================

  - fix regression in fetching current group
    [youenn]
  - poundctl restart need to use the right config file as well
    [jensens]
  - In term of portability, bash scripts now use env to get their shell.
  - The poundctl script overrides the control binary present in $prefix/sbin.
    So are introduced options to fix that problem: (kiorky)

        - A `poundcontrol` script is added to wrap the real poundctl binary.
        - Customize the names of generated scripts.
          To use them, simply add  `poundctl-binary`, `poundrun-binary`,
          `poundcontrol-binary` to your options.

0.5.1 (2008-07-08)
==================

  - fix bug in fecthing current group
    [youenn]

0.5.0 (2008-07-07)
==================

  - rename iw.recipe.pound to plone.recipe.pound
    [youenn]
  - fix doctests
    [youenn]
  - add socket option
    [mathieu pasquet]
  - config recipe now additionally creates "poundrun" script
    which merely runs configured pound executable with correct configuration
    [rockyburt]
  - fetching current running user is now more robust
    [rockyburt]


0.4.1 (2008-07-02)
==================

  - fix Cheetah require
    [youenn]

0.4.0 (2008-06-26)
==================

  - add a ctl script (as plone.recipe.squid)
    [youenn]

0.3.0
=====

  - devide recipe in two part : build and config part
    [youenn]
  - add options to build (extra-options)
    [youenn]
  - add options to the config file (time-out, priority ...)
    [youenn]

0.2.0
=====

 - added a script to launch pound with the right conf file
   [tarek]

0.1.0
=====

 - Initial release
   [tarek]


plone/recipe/pound/tests/test_pounddocs.py
# -*- coding: utf-8 -*-
# Copyright (C)2007 'ingeniweb'

# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2 of the License, or
# (at your option) any later version.

# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.

# You should have received a copy of the GNU General Public License
# along with this program; see the file COPYING. If not, write to the
# Free Software Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
"""
Generic Test case for 'plone.recipe.pound' doctest
"""
__docformat__ = 'restructuredtext'

import unittest
import doctest
import sys
import os
import zc.buildout.testing

from zope.testing import doctest, renormalizing

optionflags =  (doctest.ELLIPSIS |
                doctest.NORMALIZE_WHITESPACE |
                doctest.REPORT_ONLY_FIRST_FAILURE)
current_dir = os.path.dirname(__file__)


import urllib2

def urlopen(url):
    return open(os.path.join(current_dir, 'Pound-2.3.2.tgz'))

urllib2.urlopen = urlopen

def setUp(test):
    zc.buildout.testing.buildoutSetUp(test)

    # Install the recipe in develop mode
    zc.buildout.testing.install_develop('plone.recipe.pound', test)
    zc.buildout.testing.install_develop('zc.recipe.cmmi', test)
    zc.buildout.testing.install_develop('Cheetah', test)

def test_suite():

    globs = globals()
    globs['test_dir'] = current_dir
    globs['bin_dir'] = '%s/bin' % current_dir

    suite = unittest.TestSuite((
            doctest.DocFileSuite(
                '../docs/building.txt',
                setUp=setUp,
                tearDown=zc.buildout.testing.buildoutTearDown,
                optionflags=optionflags,
                globs = globs,
                checker=renormalizing.RENormalizing([
                        # If want to clean up the doctest output you
                        # can register additional regexp normalizers
                        # here. The format is a two-tuple with the RE
                        # as the first item and the replacement as the
                        # second item, e.g.
                        # (re.compile('my-[rR]eg[eE]ps'), 'my-regexps')
                        zc.buildout.testing.normalize_path,
                        ]),
                ),
            doctest.DocFileSuite(
                '../docs/configuring.txt',
                setUp=setUp,
                globs = globs,
                tearDown=zc.buildout.testing.buildoutTearDown,
                optionflags=optionflags,
                checker=renormalizing.RENormalizing([
                        # If want to clean up the doctest output you
                        # can register additional regexp normalizers
                        # here. The format is a two-tuple with the RE
                        # as the first item and the replacement as the
                        # second item, e.g.
                        # (re.compile('my-[rR]eg[eE]ps'), 'my-regexps')
                        zc.buildout.testing.normalize_path,
                        ]),
                ),
            doctest.DocFileSuite(
                '../docs/technical.txt',
                setUp=setUp,
                globs = globs,
                tearDown=zc.buildout.testing.buildoutTearDown,
                optionflags=optionflags,
                checker=renormalizing.RENormalizing([
                        # If want to clean up the doctest output you
                        # can register additional regexp normalizers
                        # here. The format is a two-tuple with the RE
                        # as the first item and the replacement as the
                        # second item, e.g.
                        # (re.compile('my-[rR]eg[eE]ps'), 'my-regexps')
                        zc.buildout.testing.normalize_path,
                        ]),
                ),

            ))
    return suite

if __name__ == '__main__':
    unittest.main(defaultTest='test_suite')


plone/recipe/pound/build.py
# -*- coding: utf-8 -*-
# Copyright (C)2007 'ingeniweb'

# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2 of the License, or
# (at your option) any later version.

# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.

# You should have received a copy of the GNU General Public License
# along with this program; see the file COPYING. If not, write to the
# Free Software Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
"""Build Recipe pound"""

import os
import stat
import zc
import logging

from zc.recipe.cmmi import Recipe as CMMIRecipe
from plone.recipe.pound import utils

TEMPLATE = """\
#!/bin/sh

%(pound)s -f %(cfg)s -p %(pid)s

"""


class BuildRecipe(CMMIRecipe):
    """This recipe is used by zc.buildout"""

    def __init__(self, buildout, name, options):
        CMMIRecipe.__init__(self, buildout, name, options)
        self.buildout = buildout
        self.logger=logging.getLogger(self.name)

    def install(self):
        """installer"""
        # building extra options
        owner = self.options.get('owner', utils.get_sysuser())
        group = self.options.get('group', utils.get_sysgroup())
        ssl_dir = self.options.get('ssl_dir', None)
        t_rsa = self.options.get('t_rsa', None)
        extra_options = self.options.get('extra-options', '')
        extra = ''
        if '--with-owner' not in extra_options:
            extra += ' --with-owner=%s' % owner
        if '--with-group' not in extra_options:
            extra += ' --with-group=%s' %  group
        if ssl_dir is not None:
            if os.path.isdir(ssl_dir):
                extra += ' --with-ssl=%s' % ssl_dir
            else:
                self.logger.error(
                    "You need to specify an valid directory for ssl directory")
                raise zc.buildout.UserError("ssl directory is invalid")
        if t_rsa is not None:
            try:
                t_rsa = int(t_rsa)
                extra += ' --with-t_rsa=%d' % (t_rsa,)

            except ValueError:
                self.logger.error(
                    "You need to specify an integer for timeout rsa")
                raise zc.buildout.UserError("Time out rsa is invalid")
        if extra_options is not None:
            extra += ' %s' % (extra_options,)
        self.logger.info('compilation option : %s' % (extra,))

        self.extra_options = extra

        # uses cmmi installer
        installed = CMMIRecipe.install(self)

        # building the script that launches pound
        command = os.path.join(self.options['location'], 'sbin', 'pound')
        var_dir = os.path.join(self.options['location'], 'var')
        pid = os.path.join(var_dir, 'pound.pid')

        script = TEMPLATE % {'pound': command, 'cfg': self.getFileNameConfig(),
                             'pid': pid}

        bin_dir = self.buildout['buildout']['bin-directory']
        script_name = os.path.join(bin_dir, self.name)
        f = open(script_name, 'wb')
        f.write(script)
        f.close()
        mode = os.stat(script_name).st_mode
        os.chmod(script_name, mode | stat.S_IXUSR | stat.S_IXGRP | stat.S_IXOTH)
        return (installed, script_name)


    def getFileNameConfig(self):
        return 'pound.cfg'

    def update(self):
        pass


plone/recipe/pound/__init__.py
# -*- coding: utf-8 -*-
# Copyright (C)2007 'Ingenploneeb'

# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2 of the License, or
# (at your option) any later version.

# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.

# You should have received a copy of the GNU General Public License
# along with this program; see the file COPYING. If not, write to the
# Free Software Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
"""Recipe pound"""

from build import BuildRecipe
from config import ConfigureRecipe



plone/recipe/pound/config.py
# -*- coding: utf-8 -*-
# Copyright (C)2008 'ingeniweb'

# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2 of the License, or
# (at your option) any later version.

# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.

# You should have received a copy of the GNU General Public License
# along with this program; see the file COPYING. If not, write to the
# Free Software Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
"""Config Recipe pound"""

import os
import stat
import zc
from types import BooleanType
from Cheetah.Template import Template

from plone.recipe.pound import utils

RECIPE_BUILD_NAME = 'plone.recipe.pound:build'


def get_options_from_build(buildout, options):
    part = options.get('poundbuildpart', None)
    if part:
        return buildout[part]

    for part in buildout.keys():
        if buildout[part].has_key('recipe') and \
            buildout[part]['recipe'] == RECIPE_BUILD_NAME:
            return buildout[part]
    return {}


class ConfigureRecipe(object):
    """This recipe is used by zc.buildout"""

    def __init__(self, buildout, name, options):
        self.buildout = buildout
        location = options.get(
            'location', buildout['buildout']['parts-directory'])
        options['location'] = os.path.join(location, name)
        options['prefix'] = options['location']

        self.options = options
        self.buildoptions = get_options_from_build(buildout, options)
        self.options['socket'] = self.buildoptions.get(
            'socket',
            os.path.join(
                self.buildout['buildout']['directory'],
                'var',
                'pound.sock'
            )
        )

    def install(self):
        """ install config fo pound """

        # configures pound
        # loading templates
        curdir = os.path.dirname(__file__)
        tpl = Template(open(os.path.join(curdir, 'pound.cfg.tpl')).read())
        try:
            tpl.daemon = int(self.options.get('daemon',1))
        except ValueError:
            raise zc.buildout.UserError("Deamon is invalid")
        tpl.log_facility = self.options.get('log_facility','daemon')
        try:
            tpl.log_level = int(self.options.get('log_level',1))
        except ValueError:
            raise zc.buildout.UserError("LogLevel is invalid")
        try:
            tpl.alive = int(self.options.get('alive',30))
        except ValueError:
            raise zc.buildout.UserError("Alive is invalid")
        try:
            tpl.dynscale = int(self.options.get('dynscale',0))
        except ValueError:
            raise zc.buildout.UserError("Dynscale is invalid")

        try:
            tpl.client = int(self.options.get('client',10))
        except ValueError:
            raise zc.buildout.UserError("Client is invalid")
        try:
            tpl.timeout = int(self.options.get('timeout',15))
        except ValueError:
            raise zc.buildout.UserError("Timeout is invalid")

        try:
            tpl.grace = int(self.options.get('grace',30))
        except ValueError:
            raise zc.buildout.UserError("Grace is invalid")
        try:
            tpl.owner = self.buildoptions.get('owner', utils.get_sysuser())
        except ValueError:
            raise zc.buildout.UserError("Owner is invalid")
        try:
            tpl.group = self.buildoptions.get('group', utils.get_sysgroup())
        except ValueError:
            raise zc.buildout.UserError("Group is invalid")
        ## session management
        try:
            tpl.sticky = self.options.get('sticky', 'on')
        except ValueError:
            raise zc.buildout.UserError("Sticky is invalid")
        try:
            tpl.sessiontype = self.options.get('sessiontype', 'COOKIE')
        except ValueError:
            raise zc.buildout.UserError("Sessiontype is invalid")
        try:
            tpl.sessioncookie = self.options.get('sessioncookie', '__ac')
        except ValueError:
            raise zc.buildout.UserError("Sessioncookie is invalid")
        try:
            tpl.sessiontimeout = int(self.options.get('sessiontimeout', 300))
        except ValueError:
            raise zc.buildout.UserError("SessionTimeout is invalid")

        tpl.socket = self.options.get('socket', '')

        # creating balancers
        balancer_cfg = []

        for balancer in self.options.get('balancers', '').split('\n'):

            balancer = balancer.strip()
            if balancer == '':
                continue
            balancer = balancer.split()
            name = balancer[0]
            try:
                (adress, port) = balancer[1].split(':')
            except ValueError:
                raise zc.buildout.UserError("balancer syntax is not correct %s" \
                                            % balancer)

            backends = balancer[2:]
            backends_cfg = []

            for backend in backends:
                backend = backend.split(':')
                host = backend[0]
                port_backend = backend[1]
                priority = 0
                timeout = 0
                if ',' in port_backend:
                    l =  port_backend.split(',')
                    if len(l) == 3:
                        (port_backend, priority, timeout) = l

                    elif len(l) == 2:
                        (port_backend, priority) = l
                    else:
                        raise zc.buildout.UserError("Backend configuration is invalid")
                backends_cfg.append({'host': host,
                                    'port': port_backend,
                                    'priority' : priority,
                                    'timeout': timeout})

            balancer_dict = {'name': name,
                             'port': port,
                             'adress': adress,
                             'backends': backends_cfg}

            balancer_cfg.append(balancer_dict)
        tpl.balancers = balancer_cfg


        # writing the file
        if not  os.path.exists(self.options['location']):
            os.mkdir(self.options['location'])
        etc_dir = os.path.join(self.options['location'], 'etc')
        if not os.path.exists(etc_dir):
            os.mkdir(etc_dir)
        var_dir = os.path.join(self.options['location'], 'var')
        if not os.path.exists(var_dir):
            os.mkdir(var_dir)

        filename = os.path.join(etc_dir, 'pound.cfg')
        f = open(filename, 'w')

        try:
            print >>f, tpl
        finally:
            f.close()

        if self.buildoptions:
            self.options['executable'] = os.path.join(self.buildoptions['location'],
                                                      'sbin','pound')
        else:
            if not self.options.get('executable'):
                raise zc.buildout.UserError("path to executable is required")


        pid = os.path.join(var_dir, 'pound.pid')
        scripts = []
        for x in ('poundctl', 'poundrun', 'poundcontrol'):
            ctl = Template(open(os.path.join(curdir, x+'.tpl')).read())
            ctl.poundbin = self.options['executable']
            ctl.socket = self.options.get('socket', '')
            ctl.poundcontrol = os.path.join(
                os.path.dirname(self.options['executable']),
                'poundctl'
            )
            ctl.logfilename = self.options.get('log_file', os.path.join(var_dir, 'pound.log'))
            ctl.poundcfg = filename
            ctl.poundpid = pid
            bin_dir = self.buildout['buildout']['bin-directory']
            script_name = os.path.join(bin_dir, self.options.get('%s-binary' % x, x))
            try:
                os.remove(script_name)
            except OSError:
                pass
            f = open(script_name, 'w')
            try:
                print >>f, ctl
            finally:
                f.close()
            mode = os.stat(script_name).st_mode
            os.chmod(script_name, mode | stat.S_IXUSR | stat.S_IXGRP | stat.S_IXOTH)
            scripts.append(script_name)

        return [etc_dir] + scripts
